javascript/packages/fury/lib/gen/collection.ts

```diff
@@ -0,0 +1,364 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { TypeDescription } from "../description";
+import { CodecBuilder } from "./builder";
+import { BaseSerializerGenerator, RefState } from "./serializer";
+import { CodegenRegistry } from "./router";
+import { InternalSerializerType, RefFlags, Serializer } from "../type";
+import { Scope } from "./scope";
+import Fury from "../fury";
+
+export const CollectionFlags = {
+  /** Whether track elements ref. */
+  TRACKING_REF: 0b1,
+
+  /** Whether collection has null. */
+  HAS_NULL: 0b10,
+
+  /** Whether collection elements type is not declare type. */
+  NOT_DECL_ELEMENT_TYPE: 0b100,
+
+  /** Whether collection elements type different. */
+  NOT_SAME_TYPE: 0b1000,
+};
+
+class CollectionAnySerializer {
+  constructor(private fury: Fury) {
+
+  }
+
+  protected writeElementsHeader(arr: any) {
+    let flag = 0;
+    let isSame = true;
+    let serializer: Serializer | null = null;
+    let includeNone = false;
+
+    for (const item of arr) {
+      if ((item === undefined || item === null) && !includeNone) {
+        includeNone = true;
+      }
+      if (isSame) {
+        const current = this.fury.classResolver.getSerializerByData(item);
```

Comment:
This can be cached, the hash look up and method call is costly. We can cache it, and check whether element types are same as previous element to reduce this cose

Reply:
The `getSerializerByData` function can only check the 8 base types of a JavaScript object because objects in JavaScript usually do not have a type. Therefore, it does not include a hash lookup, just an array index. 
Additionally, the`writeElementsHeader` function is used only in compatible mode, where the user declares the type by using `Type.any()` rather than a specific type like `Type.string()`.  If a specific type is declared,`readStmtSpecificType` and `writeStmtSpecificType` will generate code that does not require checking if the elements are of the same type.

Reply:
I will add a collection snapshot for you to review the generated code.